### PR TITLE
stat/distuv: implement Poisson

### PR DESCRIPTION
**Pull request by sbinet**

### CODE REVIEW

File: stat/distuv/poisson.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+// Poisson implements the Poisson distribution, a discrete probability distribution
+// that expresses the probability of a given number of events occurring in a fixed
+// interval of time and/or space.
+//
+// The poisson distribution has density function:
+//
+//  λ^k / k! e^(-λ)
+//
+// For more information, see https://en.wikipedia.org/wiki/Poisson_distribution
```
Comment (btracey) [inline stat/distuv/poisson.go]: trailing .

File: stat/distuv/poisson.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	kfac := 1
+	ix := int(math.Floor(x))
+	for i := 1; i <= ix; i++ {
+		kfac *= i
+	}
+	return mathext.GammaInc(math.Floor(x+1), p.Lambda) / float64(kfac)
+}
+
+// Entropy returns the entropy of the distribution.
```
Comment (btracey) [inline stat/distuv/poisson.go]: Remove the functions that are not implemented. When this is through, could you add a PR with the unimplemented functions?

File: stat/distuv/poisson.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	Source *rand.Rand
+}
+
+// CDF computes the value of the cumulative distribution function at x.
+func (p Poisson) CDF(x float64) float64 {
+	if x < 0 {
+		return 0
+	}
+	kfac := 1
+	ix := int(math.Floor(x))
```
Comment (btracey) [inline stat/distuv/poisson.go]: Gamma(k+1) = k! .

This should be done in log space and then exponentiated

File: stat/distuv/poisson.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// ExKurtosis returns the excess kurtosis of the distribution.
+func (p Poisson) ExKurtosis() float64 {
+	return 1 / p.Lambda
+}
+
+// LogProb computes the natural logarithm of the value of the probability
+// density function at x.
+func (p Poisson) LogProb(x float64) float64 {
```
Comment (btracey) [inline stat/distuv/poisson.go]: This doesn't look right. This smoothy increases with x, but the true PDF doesn't. I think you need a floor function in here.

File: stat/distuv/poisson.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// Prob computes the value of the probability density function at x.
+func (p Poisson) Prob(x float64) float64 {
+	return math.Exp(p.LogProb(x))
+}
+
+// Rand returns a random sample drawn from the distribution.
+func (p Poisson) Rand() float64 {
+	panic("not implemented")
```
[the remark this answers is not in the document]
Comment (btracey) [reply stat/distuv/poisson.go]: I also think it's pretty important to have random number generation in the first commit if reasonable (and I think it's reasonable here).

File: stat/distuv/poisson_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+package distuv
+
+import (
+	"testing"
+
+	"gonum.org/v1/gonum/floats"
+)
+
+func TestPoisson(t *testing.T) {
```
Comment (btracey) [inline stat/distuv/poisson_test.go]: With random numbers generated, we also can use the rest of the testing capabilities here.

File: stat/distuv/poisson.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+// NumParameters returns the number of parameters in the distribution.
+func (Poisson) NumParameters() int {
+	return 1
+}
+
+// Prob computes the value of the probability density function at x.
+func (p Poisson) Prob(x float64) float64 {
+	return math.Exp(p.LogProb(x))
+}
+
```
[the remark this answers is not in the document]
Comment (btracey) [reply stat/distuv/poisson.go]: I take it back. Python uses `cephes.pdtrik` which I cannot find anywhere.

File: stat/distuv/poisson.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbinet) [reply stat/distuv/poisson.go]: sorry, I forgot to push my changes...
here they are...

File: stat/distuv/poisson.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+	"gonum.org/v1/gonum/mathext"
+)
+
+// Poisson implements the Poisson distribution, a discrete probability distribution
+// that expresses the probability of a given number of events occurring in a fixed
+// interval of time and/or space.
+//
+// The poisson distribution has density function:
+//
```
[the remark this answers is not in the document]
Comment (sbinet) [reply stat/distuv/poisson.go]: done

File: stat/distuv/poisson.go (hunk 1 above)
Comment (sbinet) [reply stat/distuv/poisson.go]: I was worried about it being mistakenly put inside the URL.
but ok, done.

File: stat/distuv/poisson.go (hunk 2 above)
Comment (sbinet) [reply stat/distuv/poisson.go]: done.

File: stat/distuv/poisson.go (hunk 4 above)
Comment (sbinet) [reply stat/distuv/poisson.go]: done.

File: stat/distuv/poisson.go (hunk 3 above)
Comment (sbinet) [reply stat/distuv/poisson.go]: (still need to go to log-space)

File: stat/distuv/poisson.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+	Source *rand.Rand
+}
+
+// CDF computes the value of the cumulative distribution function at x.
+func (p Poisson) CDF(x float64) float64 {
+	if x < 0 {
+		return 0
+	}
+	return mathext.GammaInc(math.Floor(x+1), p.Lambda) / math.Gamma(x+1)
```
Comment (sbinet) [inline stat/distuv/poisson.go]: (Still need to go to log-space)

File: stat/distuv/poisson.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+	"math/rand"
+
+	"gonum.org/v1/gonum/mathext"
+)
+
+// Poisson implements the Poisson distribution, a discrete probability distribution
+// that expresses the probability of a given number of events occurring in a fixed
+// interval of time and/or space.
+// The poisson distribution has density function:
+//  λ^k / k! e^(-λ)
```
Comment (vladimir-ch) [inline stat/distuv/poisson.go]: Could you make this a function so that it is clear what is the variable and what is the parametr?

Comment (sbinet) [reply stat/distuv/poisson.go]: done.

File: stat/distuv/poisson.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+	return 1
+}
+
+// Prob computes the value of the probability density function at x.
+func (p Poisson) Prob(x float64) float64 {
+	return math.Exp(p.LogProb(x))
+}
+
+// Rand returns a random sample drawn from the distribution.
+func (p Poisson) Rand() float64 {
```
Comment (btracey) [inline stat/distuv/poisson.go]: I'm pretty sure this is not how we want to generate random variables. It is very expensive, with an expected  `Lambda` random numbers generated per random variable. For small lambda, I think we instead want to use the inverse transform sampling in the wiki article (better source: 4.1 of http://statweb.stanford.edu/~owen/mc/Ch-nonunifrng.pdf) . For large lambda, we definitely want to use rejection sampling with a Normal distribution. I don't know where that divide is, but I would guess it's at `Lambda > 10` or so. See page 511 of http://luc.devroye.org/chapter_ten.pdf, and note the discussion on the best choice of Delta

File: stat/distuv/poisson_test.go (hunk 6 above)
Comment (btracey) [reply stat/distuv/poisson_test.go]: Ping on the extra tests to match the other variables, i.e. `testChiSquared`

File: stat/distuv/poisson.go (hunk 11 above)
Comment (kevinczimmerman) [reply stat/distuv/poisson.go]: I believe for the Numpy implementation (https://github.com/numpy/numpy/blob/master/numpy/random/mtrand/distributions.c) the divide is `Lambda >= 10` as seen in the `rk_poisson` function.

**Comment by kczimm** [comment]
Do I need to be an existing contributor to be able to help with this implementation? I'd like to address the costly computation and expedite the merge.

**Comment by sbinet** [comment]
What you could do is take the branch where I put my work in progress pull request, add your own modifications and send your own pull request.
(I would then just abandon my pull request)

**Comment by kortschak** [comment]
@sbinet Do you know why tests don't run on your PRs?

**Comment by sbinet** [comment]
even after a rebase...

**Comment by sbinet** [comment]
I guess it makes a good argument for #271 to be the canonical one and drop this PR.

File: stat/distuv/poisson.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply stat/distuv/poisson.go]: https://github.com/scipy/scipy/blob/master/scipy/special/cdflib/cdfpoi.f

[the remark this answers is not in the document]
Comment (btracey) [reply stat/distuv/poisson.go]: I don't understand. Is there some place where `pdtrik` gets renamed to `cdfpoi`? I don't see how `cephes.pdtrik` ends up calling the function you linked.

[the remark this answers is not in the document]
Comment (kortschak) [reply stat/distuv/poisson.go]: There code generation in [scipy/special/_generate_pyx.py](https://github.com/scipy/scipy/blob/master/scipy/special/_generate_pyx.py#L252) (I think I posted that to gopherdata on slack).

**Pull request closed without merging**